phpunit/class-rest-block-renderer-controller-test.php

```diff
@@ -0,0 +1,367 @@
+			'post_title' => 'Test Post',
+		) );
+	}
+
+	/**
+	 * Delete test data after our tests run.
+	 */
+	public static function wpTearDownAfterClass() {
+		self::delete_user( self::$user_id );
+	}
+
+	/**
+	 * Set up.
+	 *
+	 * @see gutenberg_register_rest_routes()
+	 */
+	public function setUp() {
+		$this->register_test_block();
+		$this->register_post_context_test_block();
+		parent::setUp();
+	}
+
+	/**
+	 * Tear down.
+	 */
+	public function tearDown() {
+		WP_Block_Type_Registry::get_instance()->unregister( self::$block_name );
+		WP_Block_Type_Registry::get_instance()->unregister( self::$context_block_name );
+		parent::tearDown();
+	}
+
+	/**
+	 * Register test block.
+	 */
+	public function register_test_block() {
+		register_block_type( self::$block_name, array(
+			'attributes'      => array(
+				'some_string' => array(
+					'type'    => 'string',
+					'default' => 'some_default',
+				),
+				'some_int'    => array(
+					'type' => 'integer',
+				),
+				'some_array'  => array(
+					'type'  => 'array',
+					'items' => array(
+						'type' => 'integer',
+					),
+				),
+			),
+			'render_callback' => array( $this, 'render_test_block' ),
+		) );
+	}
+
+	/**
+	 * Register test block with post_id as attribute for post context test.
+	 */
+	public function register_post_context_test_block() {
+		register_block_type( self::$context_block_name, array(
+			'attributes'      => array(),
+			'render_callback' => array( $this, 'render_post_context_test_block' ),
+		) );
+	}
+
+	/**
+	 * Test render callback.
+	 *
+	 * @param array $attributes Props.
+	 * @return string Rendered attributes, which is here just JSON.
+	 */
+	public function render_test_block( $attributes ) {
+		return wp_json_encode( $attributes );
+	}
+
+	/**
+	 * Test render callback for testing post context.
+	 *
+	 * @return string
+	 */
+	public function render_post_context_test_block() {
+		return get_the_title();
+	}
+
+	/**
+	 * Check that the route was registered properly.
+	 *
+	 * @covers WP_REST_Block_Renderer_Controller::register_routes()
+	 */
+	public function test_register_routes() {
+		$dynamic_block_names = get_dynamic_block_names();
+		$this->assertContains( self::$block_name, $dynamic_block_names );
+
+		$routes = $this->server->get_routes();
+		foreach ( $dynamic_block_names as $dynamic_block_name ) {
+			$this->assertArrayHasKey( "/gutenberg/v1/block-renderer/(?P<name>$dynamic_block_name)", $routes );
+		}
+	}
+
+	/**
+	 * Test getting item without permissions.
+	 *
+	 * @covers WP_REST_Block_Renderer_Controller::get_item()
+	 */
+	public function test_get_item_without_permissions() {
+		wp_set_current_user( 0 );
+
+		$request = new WP_REST_Request( 'GET', '/gutenberg/v1/block-renderer/' . self::$block_name );
+		$request->set_param( 'context', 'edit' );
+
+		$response = $this->server->dispatch( $request );
+
+		$this->assertErrorResponse( 'gutenberg_block_cannot_read', $response, rest_authorization_required_code() );
+	}
```

Comment:
Can we include additional tests for `test_get_item_without_permissions_invalid_post()` and `test_get_item_without_permissions_cannot_edit_post()`, just to make sure we have those code paths covered?